lib/rubygems/util.rb

```diff
@@ -112,4 +112,50 @@ def self.correct_for_windows_path(path)
     end
   end
 
+  class DefaultKey
+  end
+
+  class PrivateType
+  end
+
+  require "date"
+  SAFE_MARSHAL_CLASSES = [
+    Array,
+    Date,
+    Float,
+    Hash,
+    Integer,
+    String,
+    Symbol,
+    Time,
+
+    TrueClass,
+    FalseClass,
+
+    NilClass,
+
+    Gem::Dependency,
+    Gem::NameTuple,
+    Gem::Platform,
+    Gem::Requirement,
+    Gem::Specification,
+    Gem::Version,
+
+    DefaultKey,
+    PrivateType,
+  ].freeze
+  SAFE_MARSHAL_ERROR = "Unexpected class %s present in marshaled data. Only %s are allowed."
+  SAFE_MARSHAL_PROC = proc do |object|
+    object.tap do
+      unless SAFE_MARSHAL_CLASSES.include?(object.class)
+        raise TypeError, format(SAFE_MARSHAL_ERROR, object.class, SAFE_MARSHAL_CLASSES.join(", "))
+      end
+    end
+  end
+
+  private_constant :SAFE_MARSHAL_CLASSES, :SAFE_MARSHAL_ERROR, :SAFE_MARSHAL_PROC
+
+  def self.safe_load_marshal(data)
+    Marshal.load(data, SAFE_MARSHAL_PROC)
```

Comment:
I don't think this is really safe because the Proc gets called after an object has been created.
That means its `class.allocate` has been called and also `#initialize` has been called, which could be arbitrary code and e.g. does affect `ObjectSpace.each_object` and potentially the object is even referenced by other objects already if  `#initialize` stores `self` somewhere.

Reply:
Can we build a test for the kind of "unsafeness" we are trying to guard against? Is it just preventing execution of `class.allocate` and `#initialize` on classes outside the safe list?

Reply:
> Can we build a test for the kind of "unsafeness" we are trying to guard against? Is it just preventing execution of `class.allocate` and `#initialize` on classes outside the safe list?

That's exactly what Marshal cannot do and my point above.